src/sentry/api/endpoints/rule_snooze.py

```diff
@@ -0,0 +1,90 @@
+import datetime
+
+from rest_framework import serializers, status
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry import audit_log
+from sentry.api.base import region_silo_endpoint
+from sentry.api.bases.project import ProjectAlertRulePermission, ProjectEndpoint
+from sentry.api.exceptions import ConflictError
+from sentry.api.serializers import serialize
+from sentry.incidents.models import AlertRule
+from sentry.models import Rule, RuleSnooze
+
+
+class RuleSnoozeSerializer(serializers.Serializer):
+    userId = serializers.IntegerField(required=False, allow_null=True)
+    rule = serializers.BooleanField(required=False, allow_null=True)
+    alertRule = serializers.BooleanField(required=False, allow_null=True)
+    until = serializers.DateTimeField(required=False, allow_null=True)
+
+
+@region_silo_endpoint
+class RuleSnoozeEndpoint(ProjectEndpoint):
+    permission_classes = (ProjectAlertRulePermission,)
+
+    def post(self, request: Request, project, rule_id) -> Response:
+        serializer = RuleSnoozeSerializer(data=request.data)
+        if serializer.is_valid():
+            data = serializer.validated_data
+
+            issue_alert_rule = None
+            metric_alert_rule = None
+
+            if data.get("rule"):
+                try:
+                    issue_alert_rule = Rule.objects.get(id=rule_id)
+                except Rule.DoesNotExist:
+                    raise serializers.ValidationError("Rule does not exist")
+
+            if data.get("alertRule"):
+                try:
+                    metric_alert_rule = AlertRule.objects.get(id=rule_id)
+                except AlertRule.DoesNotExist:
+                    raise serializers.ValidationError("Rule does not exist")
+
+            # don't allow editing of a rulesnooze object for a given rule and user (or no user)
+            if RuleSnooze.objects.filter(
+                user_id=data.get("userId"), rule=issue_alert_rule, alert_rule=metric_alert_rule
+            ).exists():
+                raise ConflictError("RuleSnooze already exists for this rule and scope.")
+
+            RuleSnooze.objects.create(
+                user_id=data.get("userId"),
+                owner_id=request.user.id,
+                rule=issue_alert_rule,
+                alert_rule=metric_alert_rule,
+                until=data.get("until"),
+                date_added=datetime.datetime.now(),
+            )
```

Comment:
Nit: I'd use `get_or_create` here, and then raise the error when the returned `created` is false.

Reply:
I thought about doing that but I'm not sure it'd work in all instances ... for example we decide to let only managers mute rules for everyone in the UI and that's you and me, if you initially muted it for everyone 10 days and then I am trying to mute it for everyone forever, the `owner_id` won't match. Or if `until` is different but everything else is the same, wouldn't it create a new row for both of those scenarios?

Reply:
You can pass `defaults={<attrs>}` to `get_or_create`, and just pass the core things you want to select on as kwargs